### PR TITLE
Reorganize tests and remove deprecated syntax when it isn't the thing being tested

### LINKED ISSUE #1049: Move the models that print deprecation warnings into a separate folder
By separating them in a folder we can run --print-canonical on the rest of the .stan models which reduces the number of warnings printed when compiling the entire set of models.

Also easier to remove them once the deprecated features are thrown out.

You are leading the deprecation stuff @WardBrian, so just wanted to make sure you are good with this?

### PR DESCRIPTION
**Pull request by WardBrian**
This closes #1049 and is the final part of implementing stan-dev/design-docs#43 here for now.

This does the following things to the tests:

- Moves all tests which intentionally issue warnings to `test/integration/good/warning`. The exception is the pedantic mode tests, which are in the `test/integration/cli-args` folder hierarchy
- For all tests which did not intentionally issue warnings, the warnings have been silenced. This involved selectively running `--canonicalize=deprecations` and manual updating of the tests, and is the main source of the file changes. The reason this had to be selective was to avoid removing tests which were intentionally using these features, and to update the tests in `bad/` which cannot be run through the canonicalizer due to them intentionally erroring out. Care was taken such that all the same error messages are elicted
- A few tests were reorganized or folder structure changed when it was clear that they belonged together. For example, testing pedantic mode is no longer done inside a `.ml` file but rather uses actual model files.

The exception to this was any warnings related to the ode solvers. I did not touch those, and they're not immediately on the chopping block.

After this, a warning was added to the parser for any uses of the old array syntax. This was used in part to verify that all of them had been updated outside of the desired tests.

As for reviewing this PR, I recommend only really reviewing the changes to `.expected` and `.ml` files. Any test change which didn't change the result in .expected is almost by definition good to go.

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [x] OR, no user-facing changes were made

## Release notes

The old form of declaring arrays like `real a[5]`, which has been deprecated since 2.26, now issues a warning. Users should write `array[5] real a` instead. The stanc auto-formatter does this translation for you if run.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by WardBrian** [comment]
@rok-cesnovar I'm very sorry for the size of this

**Comment by WardBrian** [comment]
@rok-cesnovar the reason this is failing tests is because `ordered_probit_log` and `ordered_probit_lpmf` don't have the same set of signatures. This is definitely a bug, right?

```
ordered_probit_log(int, real, vector) => real
ordered_probit_log(array[] int, vector, vector) => real
ordered_probit_log(array[] int, vector, array[] vector) => real
```
vs
```
ordered_probit_lpmf(int, real, vector) => real
ordered_probit_lpmf(array[] int, real, vector) => real
ordered_probit_lpmf(array[] int, real, array[] vector) => real
```

**Comment by rok-cesnovar** [comment]
> This is definitely a bug, right?

Huh, this is how it was defined in stanc2 (at least the last release we used: https://github.com/stan-dev/stan/blob/v2.22.0/src/stan/lang/function_signatures.h)

```
add("ordered_probit_log", bare_expr_type(double_type()), bare_expr_type(int_type()), bare_expr_type(double_type()), bare_expr_type(vector_type()));
add("ordered_probit_log", bare_expr_type(double_type()), bare_expr_type(bare_array_type(int_type(), 1)), bare_expr_type(vector_type()), bare_expr_type(vector_type()));
add("ordered_probit_log", bare_expr_type(double_type()), bare_expr_type(bare_array_type(int_type(), 1)), bare_expr_type(vector_type()), bare_expr_type(bare_array_type(vector_type(), 1)));

add("ordered_probit_lpmf", bare_expr_type(double_type()), bare_expr_type(int_type()), bare_expr_type(double_type()), bare_expr_type(vector_type()));
add("ordered_probit_lpmf", bare_expr_type(double_type()), bare_expr_type(bare_array_type(int_type(), 1)), bare_expr_type(double_type()), bare_expr_type(vector_type()));
add("ordered_probit_lpmf", bare_expr_type(double_type()), bare_expr_type(bare_array_type(int_type(), 1)), bare_expr_type(double_type()), bare_expr_type(bare_array_type(vector_type(), 1)));
```

**Comment by rok-cesnovar** [comment]
But given that `ordered_probit_log` is implemented just as calls to `ordered_probit_lpmf` https://github.com/stan-dev/math/blob/92075708b1d1796eb82e3b284cd11e544433518e/stan/math/prim/prob/ordered_probit_log.hpp this is definitely weird and we can just expose

```stan
ordered_probit_lpmf(array[] int, vector, vector) => real
ordered_probit_lpmf(array[] int, vector, array[] vector) => real
```
as well, given that these are used via the _log version.

File: src/frontend/Input_warnings.ml [hunk 1]
```diff
@@ -20,6 +20,16 @@ let deprecated token (pos, message) =
     |> Option.value ~default:Middle.Location_span.empty in
   add_warning span message
 
+let array_syntax ?(unsized = false) (pos1, pos2) =
+  let placement = if unsized then "a type" else "a variable name" in
+  add_warning
+    (Option.value ~default:Middle.Location_span.empty
+       (Middle.Location_span.of_positions_opt pos1 pos2) )
+    ( "Declaration of arrays by placing brackets after " ^ placement
+    ^ " is deprecated and will be removed in Stan 2.31.0. Instead use the \
```
Comment (rok-cesnovar) [inline src/frontend/Input_warnings.ml:29]: This should be 2.32.0 right?

File: src/frontend/Input_warnings.ml [hunk 2]
```diff
@@ -20,6 +20,16 @@ let deprecated token (pos, message) =
...
+       array keyword before the type; this can be changed automatically using \
```
Comment (rok-cesnovar) [inline src/frontend/Input_warnings.ml:30]: `array keyword before the type. This can be changed automatically using \`

**Review by rok-cesnovar — CHANGES_REQUESTED** [review]
Looks good. Two minor things. Other than that, if the ordered_probit fix makes the tests pass merge away. Thanks!

File: src/frontend/Input_warnings.ml (hunk 1 above)
Comment (WardBrian) [reply src/frontend/Input_warnings.ml:29]: Yep, my bad!

**Comment by WardBrian** [comment]
Once the ordered_probit changes are merged I'll re-run tests here and push those two wording changes. Hopefully that will be all this needs

**Review by rok-cesnovar — APPROVED** [review]

**Comment by rok-cesnovar** [comment]
🎉

**Pull request merged**